proposals/NNNN-DXIL-Scalarization.md

```diff
@@ -0,0 +1,90 @@
+# DXIL Scalarization
+
+* Proposal: [NNNN](NNNN-DXIL-Scalarization.md)
+* Author(s): [Farzon Lotfi](https://github.com/farzonl)
+* Sponsor: [Farzon Lotfi](https://github.com/farzonl)
+* Status: **Under Consideration**
+* Impacted Projects: Clang
+
+## Introduction
+As mentioned in [DXIL.rst](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#vectors) 
+"HLSL vectors are scalarized" and "Matrices are lowered to vectors". Therefore,
+ we need to be able to scalarize datastructures, call instructions, and vector 
+ operations like mathamatical, logical and bitcasts. The goal of this proposal 
+ is to present a solution for LLVM's pass manager.
+
+## Motivation
+Without Scalarizing the data structures and call instructions we can't generate
+ valid DXIL.
+
+ ## Background
+ In DXC Scalarization and SROA are merged in unfortnate ways. Most of that code
+ can be found here: [`SROA_Parameter_HLSL` Module pass in `ScalarReplAggregatesHLSL.cpp`](https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp#L4263). 
+The specific instructions DXC operates on are `CallInst`, `LoadInst`, 
+`StoreInst`, `GetElementPtrInst`, `AddrSpaceCastInst`, `BitCastInst`, and 
+`MemIntrinsic`. Not covered in this proposal is the handling of `Allocas`.
+That will be addressed in a later proposal covering DXIL Legalization.
+
+## Proposal
+Our goal should be legal DIXL with behavioral similarity to DXC. To that end it
+ makes the most sense to use the Scalarizer pass.
+- [Scalarizer.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/Scalarizer.cpp)
+
+The `scalarizer` pass with the `-scalarize-load-store` and 
+`-scalarize-variable-insert-extract` flags  will handle the `BitCastInst`, 
+`CallInst`, `LoadInst` and `StoreInst`, and `GetElementPtrInst` cases. The 
+`AddrSpaceCastInst` case isn't relevant for upstream LLVM.
+
+`MemIntrinsic` is the odd case out. We don't currently emit memcpy for
+the same cases we do in DXC.
+In many cases operations like memcpy and memset seem to happen in global scope.
+See Example:
+- https://hlsl.godbolt.org/z/8vx1Yee7b
+
+In my observation these cases get converted into cbuffer. Since we don't have
+cbuffer support yet we can likely hold off on this cases.
```

Comment:
Aren't we also building cbuffer support right now though?

Reply:
We are but this proposal is to onboard a generic scalarizer pass that is not going to have anything unique to Direct X. There is also some out of order problems between the two work streams. from talking with Justin that work is happening after DXIL lowering and scalarization is happening before.

 To me this seems like a case we can handle separately in another proposal. We already need to have our own special DXIL pass for global arrays to convert `[2 x <3 x float>]` to `[2 x [3 x float]]`.

Reply:
We may need to update the description for this PR to clarify the onboarding of the generic scalarizer pass.
The current description is very broad.

Reply:
your comment came before I updated the proposal. I think my updates address this concern but please confirm.